rust-runtime/aws-smithy-client/src/timeout.rs

```diff
@@ -51,8 +62,163 @@ impl Settings {
     /// Sets the read timeout
     pub fn with_read_timeout(self, read_timeout: Duration) -> Self {
         Self {
-            http_read_timeout: Some(read_timeout),
+            read_timeout: Some(read_timeout),
+            ..self
+        }
+    }
+
+    /// The configured TLS negotiation timeout
+    pub fn with_tls_negotiation_timeout(self, tls_negotiation_timeout: Duration) -> Self {
+        Self {
+            tls_negotiation_timeout: Some(tls_negotiation_timeout),
             ..self
         }
     }
 }
+
+/// A service that wraps another service, adding the ability to set a timeout for requests
+/// handled by the inner service.
+#[derive(Clone, Debug)]
+pub struct TimeoutService<InnerService> {
+    inner: InnerService,
+    duration: Option<Duration>,
+}
+
+impl<InnerService> TimeoutService<InnerService> {
+    /// Given a function that will sleep a thread and timeout duration, create a new HttpRequestTimeout
+    /// that will timeout if the inner service doesn't respond before the timeout elapses.
+    pub fn new(inner: InnerService, duration: Option<Duration>) -> Self {
+        Self { inner, duration }
+    }
+
+    /// Create a new HttpRequestTimeout that will never timeout
+    pub fn no_timeout(inner: InnerService) -> Self {
+        Self {
+            inner,
+            duration: None,
+        }
+    }
+}
+
+/// A layer that wraps services in a timeout service
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct TimeoutLayer(Option<Duration>);
+
+impl TimeoutLayer {
+    /// Create a new HttpRequestTimeoutLayer
+    pub fn new(duration: Option<Duration>) -> Self {
+        TimeoutLayer(duration)
+    }
+}
+
+impl<InnerService> Layer<InnerService> for TimeoutLayer {
+    type Service = TimeoutService<InnerService>;
+
+    fn layer(&self, inner: InnerService) -> Self::Service {
+        TimeoutService {
+            inner,
+            duration: self.0,
+        }
+    }
+}
+
+pin_project! {
+    #[non_exhaustive]
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    /// A future representing a timeout timer. Wraps a [Timeout] with extra context for error reporting
```

Comment:
I don't think this comment is correct here, probably needs to describe how it changes the error type